### PR TITLE
add kernel for flatten_op, test=develop

### PR DESCRIPTION
**Pull request by JepsonWong**
add kernel for flatten_op.

### CODE REVIEW

File: paddle/fluid/operators/flatten_op.cc [hunk 1]
```diff
@@ -58,28 +65,12 @@ class FlattenOpInferShape : public framework::InferShapeBase {
...
-    framework::AttributeMap attrs;
-    attrs["shape"] = out_dims;
-    attrs["inplace"] = false;
-    // Invoke Reshape Op
-    auto reshape_op = framework::OpRegistry::CreateOp(
-        "reshape", {{"X", {Input("X")}}, {"Shape", {}}},
-        {{"Out", {Output("Out")}}}, attrs);
-    reshape_op->Run(scope, place);
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
```
Comment (JiabinYang) [inline paddle/fluid/operators/flatten_op.cc:70]: why don't use the default `GetExpectedKernelType` ?

File: paddle/fluid/operators/flatten_op.cc [hunk 2]
```diff
@@ -126,34 +117,21 @@ Case 2:
...
-        scope.FindVar(Input("X"))->Get<framework::LoDTensor>().dims();
-    framework::AttributeMap attrs;
-    attrs["shape"] = framework::vectorize2int(in_dims);
-    attrs["inplace"] = false;
-
-    auto reshape_op = framework::OpRegistry::CreateOp(
-        "reshape", {{"X", {dout_name}}, {"Shape", {}}}, {{"Out", {dx_name}}},
-        attrs);
-    reshape_op->Run(scope, place);
+ protected:
```
Comment (JiabinYang) [inline paddle/fluid/operators/flatten_op.cc:130]: why don't use the default GetExpectedKernelType ?

File: paddle/fluid/operators/flatten_op.cc [hunk 3]
```diff
@@ -228,43 +203,27 @@ class Flatten2GradOpMaker : public framework::SingleGradOpDescMaker {
...
-    attrs["shape"] = framework::vectorize2int(x_dims);
-    attrs["inplace"] = false;
-
-    auto reshape_grad_op = framework::OpRegistry::CreateOp(
-        "reshape2_grad",
-        {{"Out@GRAD", {dout_name}}, {"Shape", {}}, {"XShape", {xshape_name}}},
-        {{"X@GRAD", {dx_name}}}, attrs);
-    reshape_grad_op->Run(scope, place);
+ protected:
+  framework::OpKernelType GetExpectedKernelType(
```
Comment (JiabinYang) [inline paddle/fluid/operators/flatten_op.cc:222]: same to fwd

File: paddle/fluid/operators/flatten_op.h [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+
+    auto out_dims = framework::make_ddim(
+        FlattenKernel<DeviceContext, T>::GetOutputShape(axes, x_dims));
+
+    out->Resize(out_dims);
+    out->mutable_data(context.GetPlace(), in->type());
+    framework::TensorCopy(
+        *in, context.GetPlace(),
+        context.template device_context<platform::DeviceContext>(), out);
+    out->Resize(out_dims);
```
Comment (baiyfbupt) [inline paddle/fluid/operators/flatten_op.h:94]: Why do we need to resize `out` twice? line 91 and line 96

File: python/paddle/fluid/tests/unittests/test_flatten2_op.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
+#   Copyright (c) 2018 PaddlePaddle Authors. All Rights Reserved.
```
Comment (baiyfbupt) [inline python/paddle/fluid/tests/unittests/test_flatten2_op.py]: 2018->2019

**Review by baiyfbupt — DISMISSED** [review]

File: paddle/fluid/operators/flatten_op.h (hunk 4 above)
Comment (JepsonWong) [reply paddle/fluid/operators/flatten_op.h:94]: done

File: python/paddle/fluid/tests/unittests/test_flatten2_op.py (hunk 5 above)
Comment (JepsonWong) [reply python/paddle/fluid/tests/unittests/test_flatten2_op.py]: done

**Review by baiyfbupt — APPROVED** [review]
LGTM

**Review by JiabinYang — APPROVED** [review]
LGTM

File: paddle/fluid/operators/flatten_op.cc (hunk 1 above)
Comment (JepsonWong) [reply paddle/fluid/operators/flatten_op.cc:70]: done

File: paddle/fluid/operators/flatten_op.cc (hunk 2 above)
Comment (JepsonWong) [reply paddle/fluid/operators/flatten_op.cc:130]: done

File: paddle/fluid/operators/flatten_op.cc (hunk 3 above)
Comment (JepsonWong) [reply paddle/fluid/operators/flatten_op.cc:222]: done

**Pull request merged**